pkg/network/ebpf/c/tracer.c

```diff
@@ -282,21 +282,21 @@ int BPF_BYPASSABLE_KPROBE(kprobe__tcp_close, struct sock *sk) {
         return 0;
     }
 
-    // check if this connection was already flushed and ensure we don't flush again
-    // upsert the timestamp to the map and delete if it already exists, flush connection otherwise
-    // skip EEXIST errors for telemetry since it is an expected error
-    __u64 timestamp = bpf_ktime_get_ns();
-    if (bpf_map_update_with_telemetry(conn_close_flushed, &t, &timestamp, BPF_NOEXIST, -EEXIST) == 0) {
-        cleanup_conn(ctx, &t, sk);
-        increment_telemetry_count(tcp_close_connection_flush);
+    // check if there is an entry in the map, meaning the connection was already flushed.
+    // we don't need to insert an entry here since we are already deleting from the ongoing_connect_pid map,
+    // so if this probe runs before tcp_done there is no way it will be able to flush this tuple anyway
+    if (bpf_map_delete_elem(&conn_close_flushed, &t) != 0) {
```

Comment:
Wouldn't the `tcp_done` probe race with this one here to clean up the connection? We will still flush an empty connection in `cleanup_conn` whether a "real" connection is flushed here.

Reply:
consider a few cases:

- `tcp_done` first, then `tcp_close`: `conn_close_flushed` map is cleaned up (**NEW**), no double flush happens
- `tcp_close` first, then `tcp_done`: `tcp_done` relies on fetching the PID from the `ongoing_connect_pid` map, which is cleaned up in `tcp_close`. So in this case, `tcp_done` will bail early and never set an entry in the `conn_close_flushed` map or flush any connection.
- `tcp_done` only: entry in `conn_close_flushed` leaks, cleaned by map cleaner. connection flushed, and entry in `ongoing_connect_pid` deleted too.
- `tcp_close` only: no entry created in `conn_close_flushed` (**NEW**), `ongoing_connect_pid` entry deleted
- both probes execute concurrently. I'm not sure much changes here with this change. Before we could have flushed a connection twice in an edge case, now that is still possible.

Reply:
The last case is what I don't think works anymore. With the concurrent update to conn_close_flushed by both probes we had a mutex of sorts going; that is not present anymore, and so you could call `cleanup_conn` twice while before you wouldn't. Or am I misreading this?

Reply:
hm okay I think you're right.. we can't get rid of this. I'm going to close this PR and open another one with just the config changes for this map and one other I see filling up consistently